lfsapi/auth.go

```diff
@@ -66,10 +66,13 @@ func (c *Client) doWithAuth(remote string, access creds.Access, req *http.Reques
 
 	res, err := c.doWithCreds(req, credWrapper, access, via)
 	if err != nil {
+		tracerx.Printf("dx: a: %s %v\n", err, errors.IsAuthError(err))
 		if errors.IsAuthError(err) {
-			newAccess := access.Upgrade(getAuthAccess(res))
+			newMode, newModes := getAuthAccess(res, access.Mode(), c.access)
+			newAccess := access.Upgrade(newMode)
```

Comment:
Looking through the code in this section I was struck by the fact that this function's caller, `DoWithAuth()`, appears to call itself recursively when the request header `Authorization` has been deleted but the server sent a 401; this is sort of a signal from the client to itself to try again, as it deletes the `Authorization` header just a bit below in this block.  That logic looks to have been introduced in #3293, along with similar blocks in the basic [download](https://github.com/git-lfs/git-lfs/blob/ee3088d4804bd2b699eb6a76aa1318d1d2283ed6/tq/basic_download.go#L280-L282) and [upload](https://github.com/git-lfs/git-lfs/blob/ee3088d4804bd2b699eb6a76aa1318d1d2283ed6/tq/basic_upload.go#L228-L239) adapters.

But I don't see a way to break out of this loop should the server keep sending 401s, and so the `getAuthAccess()` changes here which add another endless loop of access modes (negotiate -> basic -> negotiate -> basic -> ...) leaves me uncomfortable.

I tried, as an experiment, taking this PR's branch and replacing a [conditional](https://github.com/git-lfs/git-lfs/blob/ee3088d4804bd2b699eb6a76aa1318d1d2283ed6/t/cmd/lfstest-gitserver.go#L1416-L1420) in the `t/cmd/lfstest-gitserver.go` test server with an unconditional return of a 401, e.g.:
```go
	writeLFSError(w, 401, "Error: Authorization Required")
	return true
```
which simulates a broken server that keeps returning 401s.  Sure enough, the `t/t-credential.sh` test suite hangs indefinitely on a [push](https://github.com/git-lfs/git-lfs/blob/ee3088d4804bd2b699eb6a76aa1318d1d2283ed6/t/t-credentials.sh#L409) in `"credentials from lfs.url"` test.  Adding some tracing in `lfsapi/auth.go` shows it definitely just keeps calling `doWithAuth()` again, and while doing so, it keeps getting `basic` back from `getAuthAccess()`.

Personally I would love to see the logic here made somehow a bit more clear and with a definite end.  Naively, I would expect this code to start with a list of supported access modes (like this PR defines in `AllAccessModes()`) and then try those in sequence, stopping when they run out.  If the server replies with a way to negotiate, great, but I feel like there should be a simple `for` here stepping through the available modes the client supports, skipping any not supported by the server, and stopping when the list is exhausted.

I realize that adds to the scope, but I'm honestly unsure how to assess the changes here at the moment, given the recursive situation that appears to exist.

Reply:
It looks like we already have that behavior in the main branch.  What that means is that you'll get prompted for credentials again if your password is bad, since we'll erase the credentials and try again.  In our test code, we don't ever prune the credentials like a real credential helper, so we don't see that behavior.  If you enter real credentials but don't have access, you're supposed to get a 403 or 404, not a 401.

I'm going to remove the cycling behavior and preserve this behavior, because it's intended to work that way from what I can tell.

Reply:
It's definitely behaviour that's already in the main branch, so it's not immediately germane to this PR, but I couldn't untangle the two cycling interactions myself.

What I was struck by with the existing code is that if the server was broken somehow and always returned 401 instead of the expected 403, I'm not sure we wouldn't just enter an infinite loop.  I would certainly expect that if the server returned 403, we'd prompt with another chance to re-enter credentials.  But the logic in these three places (one here and one each in the upload and download adapters), where there's a recursive call without a maximum number of calls, just makes me suspicious that a bad server or a bad credential helper or something could put the client into a loop.  I'd love to be proven wrong, though.

Reply:
We have the same behaviour as Git does here.  If we get a 403, we give up, since the user has valid credentials, but they're not authorized to access the data.  That's always fatal.

We will technically enter an infinite loop if we keep getting 401s, but we'll keep prompting them for credentials using `git credential fill`, since we'll eventually always fall back to Basic.  Technically the user can just keep filling the credentials, or they can use a credential helper like ours in the test that never returns anything different, in which case they'll loop forever.  But that's also, as I understand it from working on it, Git's behaviour here as well, and in the normal case, they'll give up because the credential prompt is annoying or get a 403 or 404 and stop.  Notably, in many cases, the server will actually return a 403 or 404 if there are credentials provided and they're not valid (since that represents anonymous access, which isn't valid), which is often why this typically does abort in the general case.

Technically, we could hit that case, but, as I said, that's also true of Git, and I'm not worried about it.  I've removed my new cycling behavior, and we just devolve into an instance of Basic, so hopefully that's at least easier to reason about here.  If you want to add a check on the number of retries we do, I'm happy to do so, but I'd prefer to do so in a new PR, if that's all right with you.  If it's really too hard to reason about and it's important that it's in this PR, I can do that, though.

Reply:
No, I trust your experience here; and I don't think this should change in this PR, at a minimum.